### ISSUE TITLE
small uninitialized types are unsound

### ISSUE THREAD

**Issue by RustyYato** [Lucretiel/stowaway#8]
Similar to #3, but it can't be fixed by zeroing beforehand

```rust
struct Thing(MaybeUninit<usize>);
fn main() {
    assert_eq!(size_of::<Thing>(), size_of::<*mut Thing>());
    let thing = Thing(MaybeUninit::uninit());
    let stowed = Stowaway::new(thing); // miri fails here, encountered uninitialized bytes
}
```

**Comment by Lucretiel** [comment]
Can you create a PR with a test case that fails in miri so that I can investigate further and protect against future regressions?

**Comment by RustyYato** [comment]
Sure

**Comment by Lucretiel** [comment]
FWIW, it appears that the issue isn't that MIRI doesn't copy the uninitialized bytes, but that it doesn't consider filling a uninitialized int with them as UB until the int is "used" in some way. However, this is considered a bug in MIRI (https://github.com/rust-lang/miri/issues/1340), so the unsoundness is still unsound. I'll be publishing a PR later today and I'd love if you could review it.

**Comment by RustyYato** [comment]
Cool, I wasn't aware of that! Thanks for the link.

**Comment by Lucretiel** [comment]
There's also a lot of good discussion in the answer & comments of the [stackoverflow question](https://stackoverflow.com/questions/61114026/does-stdptrwrite-transfer-the-uninitialized-ness-of-the-bytes-it-writes) I posted on this topic, especially related to the current progress in defining the soundness rules for this sort of thing.

**Comment by Lucretiel** [comment]
Published [2.0.0](https://crates.io/crates/stowaway)
